### DISCUSSION TITLE
本机安装了 daed , mosdns 和 dnsmasq，应该如何配置 dns 和 route

### DISCUSSION THREAD

**Discussion by luckyjamesriver** [daeuniverse/dae#672] [Q&A] [answered]
我在一台Debian12 中配置了 daed 、 mosdns 、dnsmasq 和 sing-box 作为整个网络的旁路由，连接 VPS，目前连接正常。

Dns：

ipversion_prefer: 4
upstream {
  mosdns: 'tcp://127.0.0.1:5335'
}
routing {
  request {
    fallback: mosdns
    }
}

Route:

pname(mosdns) && l4proto(tcp) && dport(5335) && dport(53)-> must_direct
pname(NetworkManager, systemd-resolved,Dnsmasq) -> must_direct
pname(sing-box) -> must_direct
 dip(224.0.0.0/3) -> direct
 dip(geoip:private) -> direct

dip(geoip:private) -> direct
dip(geoip:cn) -> direct
domain(geosite:cn) -> direct

ipversion(6) -> block

domain(geosite:category-ads) -> block
 domain(geosite:category-ads-all) -> block

  domain(geosite:geolocation-!cn) -> toWorld_proxy
 
  !dip(geoip:cn) -> toWorld_proxy

fallback: toWorld_proxy

Dnsmasq: nameserver 127.0.0.1#5335

但是，有以下的困惑：

在 https://github.com/daeuniverse/dae/discussions/245 示例中，pname(mosdns) -> must_rules 这个配置；
在 https://github.com/daeuniverse/dae/blob/main/docs/en/configuration/external-dns.md 中 “本地主机上的外部 DNS：pname(AdGuardHome) && l4proto(udp) && dport(53) -> must_direct“ 这样配置。

请教，如何配置才是最佳方案？
谢谢！

**Comment by nagaame** [comment, marked as answer]
可以参考我的, MOSDNS 不开启 DNS 转发

Routing:

```
#...

pname(mosdns) && l4proto(udp) && dport(5335) -> must_direct
pname(dnsmasq) && l4proto(udp) && dport(53) -> must_direct

pname(mosdns) -> must_rules

```

DNS

```
#ipversion_prefer: 4
upstream {
  localdns:'udp://127.0.0.1:53'
  googledns: 'https://dns.google/dns-query'
  mosdns: 'tcp+udp://127.0.0.1:5335'
}
routing {
  request {
    qname(geosite:category-ads-all) -> reject
    qname(geosite:google@cn) -> mosdns
    qname(geosite:cn) -> mosdns
    qtype(a, aaaa) -> googledns
    qtype(cname) -> googledns
    qname(geosite:gfw) -> googledns
    qname(suffix: abc.com, keyword: apple) -> mosdns
    fallback:mosdns
  }
  response {
    upstream(googledns) -> accept
    upstream(mosdns) -> accept
    upstream(localdns) && ip(geoip:cn) -> accept
    ip(geoip:private) && !qname(geosite:cn) -> googledns
    fallback:googledns
  }
}

```

**Reply by ddatsh** [reply]
在routing 的reqeust里,都没有规则会走localdns吧,下面的response 里的 upstream(localdns) && ip(geoip:cn) -> accept
这是不会走到的?

**Comment by zuilintan** [comment]
踩了很多坑，总结一下我的配置：
[DAE配合MosDns的方案 #947](https://github.com/daeuniverse/dae/discussions)
